GVFS/GVFS.Common/NuGetUpgrade/ModernNuGetUpgrader.cs

```diff
@@ -0,0 +1,256 @@
+
+                return false;
+            }
+            catch (SerializationException exception)
+            {
+                message = string.Format("Parse error: could not parse response from server({0}). {1}", this.OrgInfoServerUrl, exception.Message);
+                this.TraceException(exception, nameof(this.TryQueryNewestVersion), $"Error parsing response from server.");
+
+                return false;
+            }
+
+            if (this.highestVersionAvailable != null &&
+                this.highestVersionAvailable > this.installedVersion)
+            {
+                newVersion = this.highestVersionAvailable;
+            }
+
+            if (newVersion != null)
+            {
+                this.tracer.RelatedInfo($"{nameof(this.TryQueryNewestVersion)} - new version available: installedVersion: {this.installedVersion}, highestVersionAvailable: {newVersion}");
+                message = $"New version {this.highestVersionAvailable} is available.";
+                return true;
+            }
+            else if (this.highestVersionAvailable != null)
+            {
+                this.tracer.RelatedInfo($"{nameof(this.TryQueryNewestVersion)} - up-to-date");
+                message = $"highest version available is {this.highestVersionAvailable}, you are up-to-date";
+                return true;
+            }
+            else
+            {
+                this.tracer.RelatedInfo($"{nameof(this.TryQueryNewestVersion)} - no versions available from feed.");
+                message = $"No versions available via endpoint.";
+            }
+
+            newVersion = this.highestVersionAvailable;
+            message = null;
+            return true;
+        }
+
+        public class ModernNuGetUpgraderConfig : NuGetUpgraderConfig
+        {
+            public ModernNuGetUpgraderConfig(ITracer tracer, LocalGVFSConfig localGVFSConfig)
+                : base(tracer, localGVFSConfig)
+            {
+            }
+
+            public string OrgInfoServer { get; set; }
+
+            public string UpgradeRing { get; set; }
+
+            public override bool TryLoad(out string error)
+            {
+                if (!base.TryLoad(out error))
+                {
+                    return false;
+                }
+
+                if (!this.localConfig.TryGetConfig(GVFSConstants.LocalGVFSConfig.OrgInfoServerUrl, out string orgInfoServerUrl, out error))
+                {
+                    return false;
+                }
+
+                this.OrgInfoServer = orgInfoServerUrl;
+
+                if (!this.localConfig.TryGetConfig(GVFSConstants.LocalGVFSConfig.UpgradeRing, out string upgradeRing, out error))
+                {
+                    this.tracer.RelatedError(error);
+                    return false;
+                }
+
+                this.UpgradeRing = upgradeRing;
+
+                return true;
+            }
+
+            public override bool IsReady(out string error)

```

Comment:
Can we do better with this method have have it only output the required settings that were missing.  Something like:
`public override bool IsReady(out List<string> missingSettings)` and have the caller build the error string?  Then you could have the specific fields that were missing instead of saying "One or more".